[In- review] Improving pageable extension adding autopagination

veronicagg opened the pull request:
Updating x-ms-pageable extension to add auto-pagination by default and "as_lazy" methods to iterate over pages of results.

azurecla commented:
Hi __@veronicagg__, I'm your friendly neighborhood Azure Pull Request Bot (You can call me AZPRBOT). Thanks for your contribution!
    <span>
        In order for us to evaluate and accept your PR, we ask that you sign a contribution license agreement. It's all electronic and will take just minutes. I promise there's no faxing. https://cla.azure.com.
    </span>

TTYL, AZPRBOT;

azurecla commented:
Hi __@veronicagg__, I'm your friendly neighborhood Azure Pull Request Bot (You can call me AZPRBOT). Thanks for your contribution!
    <span>
        In order for us to evaluate and accept your PR, we ask that you sign a contribution license agreement. It's all electronic and will take just minutes. I promise there's no faxing. https://cla.azure.com.
    </span>

TTYL, AZPRBOT;

vishrutshah commented on AutoRest/Generators/Ruby/Azure.Ruby/AzureRubyCodeGenerator.cs:
```diff
@@ -58,6 +66,78 @@ public override void NormalizeClientModel(ServiceClient serviceClient)
+                var method = serviceClient.Methods[i];
+                if (method.Extensions.ContainsKey(AzureExtensions.PageableExtension))
+                {
+                    var pageableExtension = JsonConvert.DeserializeObject<Model.PageableExtension>(method.Extensions[AzureExtensions.PageableExtension].ToString());
```
Can pageableExtension be null?

veronicagg replied:
Added a check

vishrutshah commented on AutoRest/Generators/Ruby/Azure.Ruby/AzureRubyCodeGenerator.cs:
```diff
@@ -58,6 +66,78 @@ public override void NormalizeClientModel(ServiceClient serviceClient)
+        {
+            if (serviceClient == null)
+            {
+                throw new ArgumentNullException("serviceClient");
```
In general my preferece'd be to use Nameof operator instead of string like throw new ArgumentNullException(nameof(serviceClient)); so that just incase we change the name of parameter it'll force us to change this line as well.

veronicagg replied:
Updated, thanks!

vishrutshah commented on AutoRest/Generators/Ruby/Azure.Ruby/AzureRubyCodeGenerator.cs:
```diff
@@ -58,6 +66,78 @@ public override void NormalizeClientModel(ServiceClient serviceClient)
+                if (method.Extensions.ContainsKey(AzureExtensions.PageableExtension))
+                {
+                    var pageableExtension = JsonConvert.DeserializeObject<Model.PageableExtension>(method.Extensions[AzureExtensions.PageableExtension].ToString());
+                    if (!method.Extensions.ContainsKey("nextLinkMethod") && !string.IsNullOrWhiteSpace(pageableExtension.NextLinkName)) 
```
minor: extra space at the end of line 91.

veronicagg replied:
fixed

vishrutshah commented on AutoRest/Generators/Ruby/Azure.Ruby/AzureRubyCodeGenerator.cs:
```diff
@@ -58,6 +66,78 @@ public override void NormalizeClientModel(ServiceClient serviceClient)
+                    if (!method.Extensions.ContainsKey("nextLinkMethod") && !string.IsNullOrWhiteSpace(pageableExtension.NextLinkName)) 
+                    {
+                        serviceClient.Methods.Insert(i, (Method)method.Clone());
+                        method.Name = CodeNamer.GetMethodName(method.Name);
```
why are you normalizing method name here? Shouldn't it be already normalized?

veronicagg replied:
corrected

vishrutshah commented on AutoRest/Generators/Ruby/Azure.Ruby/AzureRubyCodeGenerator.cs:
```diff
@@ -58,6 +66,78 @@ public override void NormalizeClientModel(ServiceClient serviceClient)
+                    {
+                        serviceClient.Methods.Insert(i, (Method)method.Clone());
+                        method.Name = CodeNamer.GetMethodName(method.Name);
+                        i++;
```
instead of using i as pointer to manipulate methods at specific location in list, why don't we just create new list of methods object with putting new methods and replacing them at the end? Or i might be mistaken on how 'i' is being used

veronicagg replied:
this way helps with ordering of the methods in the generated code, so I kept it.

vishrutshah commented on AutoRest/Generators/Ruby/Azure.Ruby/AzureRubyCodeGenerator.cs:
```diff
@@ -58,6 +66,78 @@ public override void NormalizeClientModel(ServiceClient serviceClient)
+        /// <summary>
+        /// Changes paginated method signatures to return Page type.
+        /// </summary>
+        /// <param name="serviceClient"></param>
```
<param name="serviceClient">The service client.</param>

veronicagg replied:
fixed

vishrutshah commented on AutoRest/Generators/Ruby/Azure.Ruby/AzureRubyCodeGenerator.cs:
```diff
@@ -58,6 +66,78 @@ public override void NormalizeClientModel(ServiceClient serviceClient)
+        }
+
+        /// <summary>
+        /// Adds method to use for autopagination
```
minor: period at the end

veronicagg replied:
fixed

vishrutshah commented on AutoRest/Generators/Ruby/Azure.Ruby/AzureRubyCodeGenerator.cs:
```diff
@@ -58,6 +66,78 @@ public override void NormalizeClientModel(ServiceClient serviceClient)
+                {
+                    continue;
+                }
+                
```
minor: Generally i'd do formate document (ctrl + K, ctrl + D) in VS to get rid of unnecessary spaces.

veronicagg replied:
fixed.

vishrutshah commented on AutoRest/Generators/Ruby/Azure.Ruby/TemplateModels/PageTemplateModel.cs:
```diff
@@ -0,0 +1,124 @@
+        /// Constructs mapper for the model class.
+        /// </summary>
+        /// <returns>Mapper as string for this model class.</returns>
+        public override string ConstructModelMapper()
```
Can we not overrider it as it looks the same as the bsae class

veronicagg replied:
fixed.

vishrutshah commented:
> using System;

add license statement 

---
Refers to: AutoRest/Generators/Ruby/Azure.Ruby/TemplateModels/PageTemplateModel.cs:1 in 620da21. [](commit_id = 620da212ab34195cb5cd86c083a397498266d042, deletion_comment = False)

vishrutshah commented on AutoRest/Generators/Ruby/Azure.Ruby/TemplateModels/PageTemplateModel.cs:
```diff
@@ -0,0 +1,124 @@
+        {
+            Parameter param = new Parameter
+            {
+                Name = RubyCodeNamer.RemoveInvalidCharacters(RubyCodeNamer.UnderscoreCase(NextLinkName)),
```
Should we create an extenison method in RubyCodeNamer as 'RubyCodeNamer.NormalizeName(string name)'

veronicagg replied:
fixed.

vishrutshah commented on AutoRest/Generators/Ruby/Azure.Ruby/TemplateModels/PageTemplateModel.cs:
```diff
@@ -0,0 +1,124 @@
+            {
+                string typeName = this.BaseModelType.Name;
+
+                if (this.BaseModelType.Extensions.ContainsKey(AzureExtensions.ExternalExtension) ||
```
Could you please elaborate why we may try to find external resource in MsRestAzure:: namespace as the condition also includes AzureExtensions.ExternalExtension

vishrutshah commented on AutoRest/Generators/Ruby/Azure.Ruby/TemplateModels/PageTemplateModel.cs:
```diff
@@ -0,0 +1,124 @@
+        /// <summary>
+        /// Gets the list of modules/classes which need to be included.
+        /// </summary>
+        public override List<string> Includes
```
why overrider here as looks same with the parent

vishrutshah commented on AutoRest/Generators/Ruby/Azure.Ruby/TemplateModels/PageTemplateModel.cs:
```diff
@@ -0,0 +1,124 @@
+            }
+        }
+
+        public override List<string> ClassNamespaces
```
why overrider here as looks same with the parent

vishrutshah commented on AutoRest/Generators/Ruby/Azure.Ruby/TemplateModels/PageTemplateModel.cs:
```diff
@@ -0,0 +1,124 @@
+        }
+    }
+}
+
```
minor: extra line :(

veronicagg replied:
fixed

vishrutshah commented on AutoRest/Generators/Ruby/Azure.Ruby/TemplateModels/PageTemplateModel.cs:
```diff
@@ -0,0 +1,124 @@
+using System.Collections.Generic;
+using System.Linq;
+using System.Text;
+using System.Threading.Tasks;
```
Do we need threading library?

veronicagg replied:
removed unnecessary usings.

vishrutshah commented:
>         /// Generates C# code for service client.

just notices this we probably change it to Azure.Ruby

---
Refers to: AutoRest/Generators/Ruby/Azure.Ruby/AzureRubyCodeGenerator.cs:164 in 620da21. [](commit_id = 620da212ab34195cb5cd86c083a397498266d042, deletion_comment = False)

vishrutshah commented on AutoRest/Generators/Ruby/Azure.Ruby/AzureRubyCodeGenerator.cs:
```diff
@@ -106,7 +187,22 @@ public static void CorrectFilterParameters(ServiceClient serviceClient)
             // Models
-            foreach (var model in serviceClient.ModelTypes)
+
+            // Paged Models
```
PageTemplateModel being an special type of ModelTemplateModel. can we hide details somewhere to make model writtting more general here?

vishrutshah commented on AutoRest/Generators/Ruby/Azure.Ruby/TemplateModels/AzureMethodTemplateModel.cs:
```diff
@@ -52,6 +52,76 @@ public bool IsLongRunningOperation
+        public string InvokeNextMethodAsync()
+        {
+            RubyCodeNamer cn = new RubyCodeNamer();
+            var builder = new IndentedStringBuilder();
```
Curious do you need IndentedStringBuilder or StringBuilder can work?

vishrutshah commented on AutoRest/Generators/Ruby/Azure.Ruby/TemplateModels/AzureMethodTemplateModel.cs:
```diff
@@ -52,6 +52,76 @@ public bool IsLongRunningOperation
+        }
+
+        /// <summary>
+        /// Returns generated response or body of the auto-paginated method
```
minor "." period in the end for consistency

vishrutshah commented on AutoRest/Generators/Ruby/Azure.Ruby/TemplateModels/AzureMethodTemplateModel.cs:
```diff
@@ -52,6 +52,76 @@ public bool IsLongRunningOperation
+                    if (compositeType.Extensions.ContainsKey(AzureExtensions.PageableExtension))
+                    { 
+                        bool isNextLinkMethod = this.Extensions.ContainsKey("nextLinkMethod") && (bool)this.Extensions["nextLinkMethod"];
+                        var isPageable = compositeType.Extensions[AzureExtensions.PageableExtension];
```
should we typercase to bool here?

vishrutshah commented on AutoRest/Generators/Ruby/Azure.Ruby/Templates/AzureMethodTemplate.cshtml:
```diff
@@ -1,19 +1,18 @@
 {
 @:@(Include(new MethodTemplate(), (MethodTemplateModel)Model))
 }
-else if (Model.HttpMethod == HttpMethod.Post || Model.HttpMethod == HttpMethod.Delete)
```
curious how come we don't need differenciator between (Post || Delete) and other? Did we change something or it was dead code?

veronicagg replied:
Due to this change [patch support for long running operations](https://github.com/Azure/autorest/commit/e1d41ab6ba2f62ceadcc6716675ed25662fb6eeb)

vishrutshah commented on AutoRest/Generators/Ruby/Azure.Ruby/Templates/AzureMethodTemplate.cshtml:
```diff
@@ -23,33 +22,23 @@ else if (Model.HttpMethod == HttpMethod.Post || Model.HttpMethod == HttpMethod.D
+def @(Model.Name)_as_lazy(@(Model.MethodParameterDeclaration))
+    response = @(Model.Name)_async(@(Model.MethodParameterInvocation)).value!
+    unless response.nil?
+        page = response.body
```
shoud it be two space seperator between block code?

veronicagg replied:
fixed.

vishrutshah commented on AutoRest/Generators/Ruby/Azure.Ruby/Templates/AzureMethodTemplate.cshtml:
```diff
@@ -23,33 +22,23 @@ else if (Model.HttpMethod == HttpMethod.Post || Model.HttpMethod == HttpMethod.D
+        page.next_method = Proc.new do |next_link|
+          @(Model.InvokeNextMethodAsync())
+        end
+    page
```
should it be two spaced more?

veronicagg replied:
fixed.

vishrutshah commented on AutoRest/Generators/Ruby/Azure.Ruby.Tests/RspecTests/paging_spec.rb:
```diff
@@ -19,102 +19,122 @@
-    result = @client.paging.get_single_pages_async().value!
-    expect(result.response.status).to eq(200)
-    expect(result.body.next_link).to be_nil
+    all_items = @client.paging.get_single_pages()
```
should we be also verifying the type of the result (all_items is of type page ) ?

veronicagg replied:
Added.

vishrutshah commented on AutoRest/Generators/Ruby/Azure.Ruby.Tests/RspecTests/paging_spec.rb:
```diff
@@ -19,102 +19,122 @@
-      count += 1
-    end
+    all_items = @client.paging.get_multiple_pages() #returns items from all the pages
+    items_count = all_items.count
```
here as well

veronicagg replied:
Updated.

vishrutshah commented on AutoRest/Generators/Ruby/Azure.Ruby.Tests/RspecTests/paging_spec.rb:
```diff
@@ -19,102 +19,122 @@
 
-    expect(count).to eq(10)
+  it 'should get multiple pages - lazy' do
+    page = @client.paging.get_multiple_pages_as_lazy()
```
same here

vishrutshah commented on AutoRest/Generators/Ruby/Ruby/TemplateModels/MethodTemplateModel.cs:
```diff
@@ -500,6 +500,24 @@ public virtual string BuildInputParameterMappings()
+            builder.AppendLine("response = {0}_async({1}).value!", Name, MethodParameterInvocation);
+            if (ReturnType.Body != null)
+            {
+                builder.AppendLine("response.body unless response.nil?"); ;
```
extra  ;

veronicagg replied:
fixed.

devigned commented:
@veronicagg please regenerate the samples. It would be useful to see how this PR changes the generated sample code.

veronicagg commented:
@devigned Current samples don't cover pagination operations, so they did not change after I regenerated with this PR.

vishrutshah commented:
> using System;

It's funny we license some files under MIT License (Ex. AzureMethodTemplateModel.cs) and some with Apache license. I am not sure which one we should but we should ask this to AutoRest team. 

---
In reply to: [229831522](https://github.com/Azure/autorest/pull/1211#issuecomment-229831522) [](ancestors = 229831522)

---
Refers to: AutoRest/Generators/Ruby/Azure.Ruby/TemplateModels/PageTemplateModel.cs:1 in 620da21. [](commit_id = 620da212ab34195cb5cd86c083a397498266d042, deletion_comment = False)

vishrutshah commented on AutoRest/Generators/Ruby/Azure.Ruby/AzureRubyCodeGenerator.cs:
```diff
@@ -58,6 +66,76 @@ public override void NormalizeClientModel(ServiceClient serviceClient)
+                        serviceClient.Methods.Insert(i, (Method)method.Clone());
+                        i++;
+                    }
+                    serviceClient.Methods[i].Extensions.Remove(AzureExtensions.PageableExtension);
```
should this line (serviceClient.Methods[i].Extensions.Remove(AzureExtensions.PageableExtension);) be inside if condition?

vishrutshah commented on AutoRest/Generators/Ruby/Azure.Ruby/TemplateModels/AzureMethodTemplateModel.cs:
```diff
@@ -52,6 +55,89 @@ public bool IsLongRunningOperation
+            var origMethodGroupedParameters = Parameters.Where(p => p.Name.Contains(Name));
+            if (origMethodGroupedParameters.Count() > 0)
+            {
+                for (int j = 0; j < nextMethod.Parameters.Count; j++)
```
in general i'd prefer foreach if i can but it's upto you.

vishrutshah commented on AutoRest/Generators/Ruby/Azure.Ruby/TemplateModels/AzureMethodTemplateModel.cs:
```diff
@@ -52,6 +55,89 @@ public bool IsLongRunningOperation
+
+                    if (param.Name.Contains(nextMethod.Name) && (param.Name.Length > nextMethod.Name.Length)) //parameter that contains the method name + postfix, it's a grouped param
+                    {
+                        string argumentName = param.Name.Replace(nextMethodName, Name);
```
should we explain bit in details, why are we changing name of the parameter on the fly...

vishrutshah commented on AutoRest/Generators/Ruby/Azure.Ruby/TemplateModels/AzureMethodTemplateModel.cs:
```diff
@@ -52,6 +55,89 @@ public bool IsLongRunningOperation
+                }
+            }
+
+            var headerParams = nextMethod.Parameters.Where(p => (p.Location == ParameterLocation.Header || p.Location == ParameterLocation.None) && !p.IsConstant && p.ClientProperty == null).Select(p => p.Name).ToList() ;
```
minor: extra space before ;

vishrutshah commented on AutoRest/Generators/Ruby/Azure.Ruby/TemplateModels/AzureMethodTemplateModel.cs:
```diff
@@ -52,6 +55,89 @@ public bool IsLongRunningOperation
+
+            var headerParams = nextMethod.Parameters.Where(p => (p.Location == ParameterLocation.Header || p.Location == ParameterLocation.None) && !p.IsConstant && p.ClientProperty == null).Select(p => p.Name).ToList() ;
+            headerParams.Add("custom_headers");
+            string nextMethodParamaterInvocation = string.Join(", ",headerParams);
```
minor: add one space in string.join just before headerParams and ,

vishrutshah commented on AutoRest/Generators/Ruby/Azure.Ruby/TemplateModels/AzureMethodTemplateModel.cs:
```diff
@@ -180,5 +266,20 @@ public override string OperationExceptionTypeString
+            {
+                if (Extensions.ContainsKey("nextMethodName") && !Extensions.ContainsKey(AzureExtensions.PageableExtension))
+                {
+                    return "Array";
```
can we be more specific like Array[Type] to help in intellisense ?

vishrutshah commented on AutoRest/Generators/Ruby/Azure.Ruby/TemplateModels/AzureServiceClientTemplateModel.cs:
```diff
@@ -18,14 +18,19 @@ public class AzureServiceClientTemplateModel : ServiceClientTemplateModel
+            PageTemplateModels = pageModels;
         }
 
+        public IList<PageTemplateModel> PageTemplateModels { get; }
```
I am not sure are we using this or not?

vishrutshah commented on AutoRest/Generators/Ruby/Azure.Ruby/Templates/AzureMethodTemplate.cshtml:
```diff
@@ -23,33 +22,22 @@ else if (Model.HttpMethod == HttpMethod.Post || Model.HttpMethod == HttpMethod.D
 @:#
 }
-@foreach (var parameter in Model.LocalParameters)
+@foreach (var parameter in Model.Parameters)
```
LocalParameters? double check

vishrutshah commented:
> @foreach (var parameter in Model.Parameters)

LocalParameters?

---
Refers to: AutoRest/Generators/Ruby/Azure.Ruby/Templates/AzureMethodTemplate.cshtml:58 in 02b714e. [](commit_id = 02b714e365598fdedb4bc81402dc2b1c3ad9eb7e, deletion_comment = False)

vishrutshah commented on AutoRest/Generators/Ruby/Azure.Ruby/Templates/PageModelTemplate.cshtml:
```diff
@@ -0,0 +1,94 @@
+# encoding: utf-8
+@Header("# ")
+@EmptyLine
+module @(Settings.Namespace)
```
double check the formatting of the model...

veronicagg commented:
> using System;

Once we confirm the license we should include, I'll update it.

---
In reply to: [230904731](https://github.com/Azure/autorest/pull/1211#issuecomment-230904731) [](ancestors = 230904731,229831522)

---
Refers to: AutoRest/Generators/Ruby/Azure.Ruby/TemplateModels/PageTemplateModel.cs:1 in 620da21. [](commit_id = 620da212ab34195cb5cd86c083a397498266d042, deletion_comment = False)

veronicagg commented:
Hi @vishrutshah, thanks for all the comments! I've addressed your feedback, feel free to take a look when you get a chance. I'm going to work on pulling from master and resolving conflicts.

vishrutshah commented on AutoRest/Generators/Ruby/Azure.Ruby/TemplateModels/AzureMethodTemplateModel.cs:
```diff
@@ -180,5 +265,21 @@ public override string OperationExceptionTypeString
+                if (Extensions.ContainsKey("nextMethodName") && !Extensions.ContainsKey(AzureExtensions.PageableExtension))
+                {
+                    SequenceType sequenceType = ((CompositeType)ReturnType.Body).Properties.Select(p => p.Type).FirstOrDefault(t => t is SequenceType) as SequenceType;
+                    return string.Format(CultureInfo.InvariantCulture, "Array<{0}>", sequenceType.ElementType.Name) ;
```
minor: extra one space just before ;

vishrutshah commented on AutoRest/Generators/Ruby/Azure.Ruby/Templates/AzureMethodTemplate.cshtml:
```diff
@@ -27,29 +26,19 @@ else if (Model.HttpMethod == HttpMethod.Post || Model.HttpMethod == HttpMethod.D
 }
-@WrapComment("# ", "@return [Concurrent::Promise] promise which provides async access to http response.")
+@WrapComment("# ", string.Format("@param custom_headers [{0}] A hash of custom headers that will be added to the HTTP request.", "Hash{String => String}"))
 #
```
why do we need to manually specify 'custom_headers' parameter? Isn't it already in LocalParameters?

veronicagg replied:
@vishrutshah They weren't in local parameters and looking into [MethodTemplate](https://github.com/Azure/autorest/blob/master/src/generator/AutoRest.Ruby/Templates/MethodTemplate.cshtml), I saw that they were added in the template code.

vishrutshah commented on AutoRest/Generators/Ruby/Azure.Ruby/Templates/AzureMethodTemplate.cshtml:
```diff
@@ -67,10 +56,11 @@ else
 @:@WrapComment("# ", string.Format("@param {0} {1}{2}", parameter.Name, parameter.Type.GetYardDocumentation(), parameter.Documentation))
 }
+@WrapComment("# ", string.Format("@param custom_headers [{0}] A hash of custom headers that will be added to the HTTP request.", "Hash{String => String}"))
 #
```
same here?

vishrutshah commented on AutoRest/Generators/Ruby/Azure.Ruby/Templates/PageModelTemplate.cshtml:
```diff
@@ -0,0 +1,94 @@
+        items
+      end
+
+      #
```
we should add  @EmptyLine at line 66, so that we have empty line between get_all_items and get_next_page methods

vishrutshah commented on AutoRest/Generators/Ruby/Azure.Ruby/Templates/PageModelTemplate.cshtml:
```diff
@@ -0,0 +1,94 @@
+      @EmptyLine
+      @:
+    }
+    
```
should we add comment for what is next_method accessor?

vishrutshah commented on AutoRest/Generators/Ruby/Azure.Ruby/Templates/PageModelTemplate.cshtml:
```diff
@@ -0,0 +1,94 @@
+      @EmptyLine
+
+      #
+      # Gets the rest of the items for the request, enabling auto-pagination 
```
. at the end of comment sentence

vishrutshah commented on AutoRest/Generators/Ruby/Azure.Ruby/Templates/PageModelTemplate.cshtml:
```diff
@@ -0,0 +1,94 @@
+      end
+
+      #
+      # Gets the next page of results
```
. at the end of comment sentence

vishrutshah commented on AutoRest/Generators/Ruby/Azure.Ruby/TemplateModels/PageTemplateModel.cs:
```diff
@@ -0,0 +1,64 @@
+        public Parameter BuildNextLinkParameter()
+        {
+            RubyCodeNamer rubyCodeNamer = new RubyCodeNamer();
+            Parameter param = new Parameter
```
I am not sure but i think we should not new RubyCodeNamer. Is there any chance we can get the normalized NextLinkName from the AzureRubyCodeGenerator class itself? what do you think about that?

vishrutshah commented on AutoRest/Generators/Ruby/Azure.Ruby/TemplateModels/AzureMethodTemplateModel.cs:
```diff
@@ -52,6 +55,88 @@ public bool IsLongRunningOperation
+        public string InvokeNextMethodAsync()
+        {
+            RubyCodeNamer cn = new RubyCodeNamer();
+            StringBuilder builder = new StringBuilder();
```
same doubt here as well. what do you think about this?

vishrutshah commented on AutoRest/Generators/Ruby/Ruby/TemplateModels/MethodTemplateModel.cs:
```diff
@@ -500,6 +511,24 @@ public virtual string BuildInputParameterMappings()
+        /// </summary>
+        public virtual string ResponseGeneration()
+        {
+            var builder = new IndentedStringBuilder("");
```
minor: IndentedStringBuilder

vishrutshah commented:
I noticed that with the latest changes autorest\Samples\azure-storage\azure.ruby\azure_storage\storage_accounts.rb got updated. we should regen samples.

vishrutshah commented:
@veronicagg almost all the changed looks good to me. I've closed addressed comments, left couple of new questions with few minor comments. Thanks!

vishrutshah commented on AutoRest/Generators/Ruby/Azure.Ruby/Templates/PageModelTemplate.cshtml:
```diff
@@ -0,0 +1,94 @@
+    @if (Model.IsPolymorphic && Model.BaseModelType == null)
+    {
+      @:@@@@discriminatorMap = Hash.new
+      @:@@@@discriminatorMap["@Model.SerializedName"] = "@Model.Name.ToLowerInvariant()"
```
value should be @Model.Name instead of @Model.Name.ToLowerInvariant(). This is regarding the bug i found. PR is [https://github.com/Azure/autorest/pull/1245](https://github.com/Azure/autorest/pull/1245)

vishrutshah commented on AutoRest/Generators/Ruby/Azure.Ruby/Templates/PageModelTemplate.cshtml:
```diff
@@ -0,0 +1,94 @@
+      @:@@@@discriminatorMap["@Model.SerializedName"] = "@Model.Name.ToLowerInvariant()"
+      foreach (var derivedType in Model.DerivedTypes)
+        {
+          @:@@@@discriminatorMap["@derivedType.SerializedName"] = "@derivedType.Name.ToLowerInvariant()"
```
same for here

vishrutshah commented:
:shipit:

vishrutshah commented:
@veronicagg Changes looks good to me!! Thanks.

veronicagg commented:
@vishrutshah great, thanks for the review!!
